checker/src/main/java/org/checkerframework/checker/guieffect/GuiEffectVisitor.java

```diff
@@ -222,20 +225,33 @@ private void lambdaAssignmentCheck(
     @Override
     public Void visitVariable(VariableTree node, Void p) {
         Void result = super.visitVariable(node, p);
-        if (node.getInitializer() != null
-                && node.getInitializer().getKind() == Tree.Kind.LAMBDA_EXPRESSION) {
-            lambdaAssignmentCheck(
-                    atypeFactory.getAnnotatedType(node),
-                    (LambdaExpressionTree) node.getInitializer(),
-                    "assignment.type.incompatible");
+        if (node.getInitializer() != null) {
+            if (node.getInitializer().getKind() == Tree.Kind.NEW_CLASS) {
+                commonAssignmentCheck(
+                        atypeFactory.getAnnotatedType(node),
+                        atypeFactory.getAnnotatedType(node.getInitializer()),
+                        node.getInitializer(),
+                        "assignment.type.incompatible");
+            } else if (node.getInitializer().getKind() == Tree.Kind.LAMBDA_EXPRESSION) {
+                lambdaAssignmentCheck(
```

Comment:
I also don't understand why that previous additional check is needed. Maybe some javadoc could explain this?

Reply:
See comment above. Can definitely add a javadoc or an inline comment block if the explanation above makes sense to you.